pandas-stubs/core/frame.pyi

```diff
@@ -577,7 +577,9 @@ class DataFrame(NDFrame, OpsMixin):
     @overload
     def set_index(
         self,
-        keys: Union[Label, Sequence],
+        keys: Union[
+            Label, Series, Index, np.ndarray, Iterator[Hashable], List[Hashable]
```

Comment:
I think `Iterator[Hashable]` includes `List[Hashable]`, but I didn't test that.

Reply:
> I think `Iterator[Hashable]` includes `List[Hashable]`, but I didn't test that.

No, it's not.  Here's an example where `pyright` shows the incompatibility:
```python
from typing import Iterator


def fun(a: Iterator[int]):
    k = next(a, None)
    while k is not None:
        print(k)


li = [1, 2, 3]
fun(li)
```
On the last line, `pyright` reports:
```text
iterlist.py:11:5 - error: Argument of type "list[int]" cannot be assigned to parameter "a" of type "Iterator[int]" in function "fun"
    "list[int]" is incompatible with protocol "Iterator[int]"
      "__next__" is not present (reportGeneralTypeIssues)
```

Reply:
`Iterable[int]` wouldn't require `__next__` but I'm not sure whether `set_index` works with any iterable.

Reply:
> `Iterable[int]` wouldn't require `__next__` but I'm not sure whether `set_index` works with any iterable.

pandas has no type annotations for `keys`. Let's keep it as-is.

Reply:
If you believe the docs, then the types that could be used are

```
Label
Series
Index
np.ndarray
Iterator
List[Union[Label, Series, Index, np.ndarray, Iterator]]
```

Clearly it also supports `Sequence` everywhere `Series` is used.